File: py/query.py
#!/usr/bin/python
# coding=utf-8

# This file contains the query for 
# - Insert Professor
# - Delete Professor
# - Modify Professor

# - Insert Discipline
# - Delete Discipline
# - Modify Discipline

# - Insert ClassRoom
# - Delete ClassRoom
# - Modify ClassRoom

# - Insert Preference
# - Delete Preference
# - Modify Preference


from SPARQLWrapper import SPARQLWrapper
import rdflib

g = rdflib.Graph()
# Set the end-point and ask if it give me the result to JSON
sparql = SPARQLWrapper("http://localhost:3030/ds/update",returnFormat="json")
sparqlQuery = SPARQLWrapper("http://localhost:3030/ds/query",returnFormat="json")


# ======================================================================
# Insert Professor
# @param
#   - id_professor, 
#   - name, 
#   - surname, 
#   - role
# ======================================================================
def insertProfessor(id_professor, name, surname, role):
    graph_professor = "http://www.rdcproject.com/graph/professor"
    
    # Create a new Query
    query = '''
            PREFIX uni: <http://www.rdfproject.com/>
            PREFIX un: <http://www.w3.org/2007/ont/unit#>
            INSERT DATA
            { 
            GRAPH <'''+graph_professor+'''>{
            uni:'''+id_professor+''' a uni:Teacher;
                                    uni:firstName "'''+name+'''"; 
                                    uni:lastName "'''+surname+'''";
                                    uni:idProfessor "'''+id_professor+'''";
                                    uni:role "'''+role+'''".
                }
            }
            '''
    
    # Run the query 
    sparql.setQuery(query)
    sparql.setMethod('POST') 
    print query
    sparql.query()

# ======================================================================
# Delete Professor
# @param
#   - id_professor
# ======================================================================
def cancelProfessor(id_professor):   
    print("ID PROFESSOR " + id_professor) 
    query = '''
                PREFIX uni: <http://www.rdfproject.com/>
                PREFIX un: <http://www.w3.org/2007/ont/unit#>
                DELETE WHERE { 
                GRAPH <http://www.rdcproject.com/graph/professor> {
                        ?object uni:idProfessor "'''+id_professor+'''";
                         ?property  ?value 
                }
                GRAPH <http://www.rdcproject.com/graph/disciplines> 
                {
                    ?objectData uni:isTaughtBy ?object. 
                }
                }
                '''

    sparql.setQuery(query)
    sparql.setMethod('POST') 
    print query
    sparql.query()

# ======================================================================
# Modify Professor
# @param
#   - id_professor, 
#   - name, 
#   - surname, 
#   - role
# ======================================================================
def modifyProfessor(id_professor, name, surname, role):   
    query = '''
            PREFIX uni: <http://www.rdfproject.com/>
            PREFIX un: <http://www.w3.org/2007/ont/unit#>
            WITH <http://www.rdcproject.com/graph/professor>
            DELETE { 
            ?x a uni:Teacher;
                uni:firstName ?oldfirstName;
                uni:lastName ?oldlastName;
                uni:role ?oldrole;
            }
            INSERT {  
            ?x a uni:Teacher;
                uni:firstName "'''+name+'''";
                uni:lastName "'''+surname+'''";
                uni:role "'''+role+'''";
            }
            WHERE { 
            ?x a uni:Teacher;
                uni:idProfessor "'''+id_professor+'''";
                OPTIONAL {
                    ?x a uni:Teacher;
                        uni:firstName ?oldfirstName;
                        uni:lastName ?oldlastName;
                        uni:role ?oldrole;
                }
            }
            '''

    sparql.setQuery(query)
    sparql.setMethod('POST') 
    print query
    sparql.query()

# ======================================================================
# Get all Professor objects
# ======================================================================
def getAllProfessors():
    query = '''
                PREFIX uni: <http://www.rdfproject.com/>
                    PREFIX un: <http://www.w3.org/2007/ont/unit#>

                    SELECT ?name ?surname ?id
                    FROM <http://www.rdcproject.com/graph/professor>
                    WHERE
                    { ?x  a uni:Teacher.
                        ?x uni:firstName ?name.
                        ?x uni:lastName ?surname.
                        ?x uni:idProfessor ?id.
                        }
                    ORDER BY ?surname
                '''

    sparqlQuery.setQuery(query)
    sparqlQuery.setMethod('POST') 
    print query
    return sparqlQuery.query().convert()



# ======================================================================
# Insert Discipline
# @param
#   - id_discipline
#   - discipline_name
#   - semester
#   - obligatory
#   - totalHours
#   - weeksHours
#   - cfu
#   - year
#   - course
#   - teacher
# ======================================================================
def insertDiscipline(id_discipline, discipline_abb, discipline_name,semester,obligatory, totalHours, weeksHours, cfu, year, curriculum, course, students, teacher):

    isTaughtBy = ""
    for t in teacher:          
        isTaughtBy += "uni:isTaughtBy uni:" + t + ";"

    
    graph_disciplines = "http://www.rdcproject.com/graph/disciplines"
    # Create a new Query
    query = '''
    PREFIX uni: <http://www.rdfproject.com/>
    PREFIX un: <http://www.w3.org/2007/ont/unit#>
    INSERT DATA
    { 
    GRAPH <'''+graph_disciplines+'''>{
    uni:'''+id_discipline +''' a uni:Discipline;
                            uni:disciplineAbbreviation "'''+discipline_abb+'''"; 
                            uni:disciplinename "'''+discipline_name+'''"; 
                            uni:semester "'''+semester+'''"; 
                            uni:obligatory "'''+obligatory+'''"; 
                            uni:totalhours "'''+totalHours+'''";
                            uni:weekhours "'''+weeksHours+'''";
                            uni:cfu "'''+cfu+'''";
                            uni:year "'''+year+'''";
                            uni:curriculum "'''+curriculum+'''";
                            uni:idDiscipline "'''+id_discipline+'''";
                            uni:numStudents "'''+students+'''";
                            '''+isTaughtBy+'''
                            uni:hasCourseof uni:'''+course+'''.
                            
        }
    }
    '''
        
    # Run the query and print the result
    sparql.setQuery(query)
    sparql.setMethod('POST') 
    print query
    sparql.query()

# ======================================================================
# Delete Discipline
# @param
#   - id_discipline
# ======================================================================
def cancelDiscipline(id_discipline):   
    query = '''
                PREFIX uni: <http://www.rdfproject.com/>
                PREFIX un: <http://www.w3.org/2007/ont/unit#>
                DELETE WHERE { 
                GRAPH <http://www.rdcproject.com/graph/disciplines> {
                        ?object uni:idDiscipline "'''+id_discipline+'''";
                         ?property  ?value 
                }
                }
                '''

    sparql.setQuery(query)
    sparql.setMethod('POST') 
    print query
    sparql.query()

# ======================================================================
# Modify Discipline
# @param
#   - id_discipline
#   - discipline_name
#   - semester
#   - obligatory
#   - totalHours
#   - weeksHours
#   - cfu
#   - year
#   - course
#   - curriculum
#   - students
#   - teacher
# ======================================================================
def modifyDiscipline(id_discipline, discipline_abb, discipline_name, semester,obligatory, totalHours, weeksHours, cfu, year, course, curriculum, students, teacher):   
    isTaughtBy = ""
    for t in teacher:          
        isTaughtBy += "uni:isTaughtBy uni:" + t + ";"

    query = '''
            PREFIX uni: <http://www.rdfproject.com/>
            PREFIX un: <http://www.w3.org/2007/ont/unit#>
            WITH <http://www.rdcproject.com/graph/disciplines>
            DELETE { 
            ?x a uni:Discipline;
                uni:idDiscipline ?oldidDiscipline;
                uni:disciplineAbbreviation ?olddisciplineAbb;
                uni:disciplinename ?olddisciplinename;
                uni:semester ?oldsemester;
                uni:obligatory ?oldobligatory;
                uni:totalhours ?oldtotalhours;
                uni:weekhours ?oldweekhours;
                uni:cfu ?oldcfu;
                uni:year ?oldwyear;
                uni:curriculum ?oldcurriculum;
                uni:numStudents ?oldnumStudents;
                uni:isTaughtBy ?oldisTaughtBy;
                uni:hasCourseof ?oldhasCourseof;
            }
            INSERT {  
            ?x a uni:Discipline;
                uni:idDiscipline "'''+id_discipline+'''";
                uni:disciplineAbbreviation "'''+discipline_abb+'''"; 
                uni:disciplinename "'''+discipline_name+'''";
                uni:semester "'''+semester+'''";
                uni:obligatory "'''+obligatory+'''";
                uni:totalhours "'''+totalHours+'''";
                uni:weekhours "'''+weeksHours+'''";
                uni:cfu "'''+cfu+'''";
                uni:year "'''+year+'''";
                uni:curriculum "'''+curriculum+'''";
                uni:numStudents "'''+students+'''";
                '''+isTaughtBy+''';
                uni:hasCourseof uni:'''+course+''';
            }
            WHERE { 
            ?x a uni:Discipline;
                uni:idDiscipline "'''+id_discipline+'''";
                OPTIONAL {
                    ?x a uni:Discipline;
                        uni:idDiscipline ?oldidDiscipline;
                        uni:disciplineAbbreviation ?olddisciplineAbb;
                        uni:disciplinename ?olddisciplinename;
                        uni:semester ?oldsemester;
                        uni:obligatory ?oldobligatory;
                        uni:totalhours ?oldtotalhours;
                        uni:weekhours ?oldweekhours;
                        uni:cfu ?oldcfu;
                        uni:year ?oldwyear;
                        uni:curriculum ?oldcurriculum;
                        uni:numStudents ?oldnumStudents;
                        uni:isTaughtBy ?oldisTaughtBy;
                        uni:hasCourseof ?oldhasCourseof;
                }
            }
            '''

    sparql.setQuery(query)
    sparql.setMethod('POST') 
    print query
    sparql.query()




# ======================================================================
# Insert ClassRoom
# @param
#   - id_room
#   - className
#   - address
#   - capacity
#   - blackboard
#   - wired
# ======================================================================
def insertClassRoom(id_room,className, address, capacity, blackboard, wired):
    graph_classrooms = "http://www.rdcproject.com/graph/classrooms" 

    # Create a new Query
    query = '''
            PREFIX uni: <http://www.rdfproject.com/>
            PREFIX un: <http://www.w3.org/2007/ont/unit#>
            INSERT DATA
            { 
            GRAPH <'''+graph_classrooms+'''>{
            uni:'''+id_room+''' a uni:Classroom;
                                    uni:classroomname "'''+className+'''"; 
                                    uni:address "'''+address+'''";
                                    uni:capacity "'''+capacity+'''";
                                    uni:blackboard "'''+blackboard+'''";
                                    uni:idRoom "'''+id_room+'''"; 
                                    uni:wired "'''+wired+'''".
                }
            }
            '''
    
    # Run the query 
    sparql.setQuery(query)
    sparql.setMethod('POST') 
    print query
    sparql.query()

# ======================================================================
# Delete ClassRoom
# @param
#   - id_room
# ======================================================================
def cancelClassRoom(id_room):   
    query = '''
                PREFIX uni: <http://www.rdfproject.com/>
                PREFIX un: <http://www.w3.org/2007/ont/unit#>
                DELETE WHERE { 
                GRAPH <http://www.rdcproject.com/graph/classrooms> {
                        ?object uni:idRoom "'''+id_room+'''";
                         ?property  ?value 
                }
                }
                '''

    sparql.setQuery(query)
    sparql.setMethod('POST') 
    print query
    sparql.query()

# ======================================================================
# Modify ClassRoom
# @param
#   - id_room, 
#   - name_room, 
#   - capacity_room, 
#   - wired_room,
#   - blackboard,
#   - address_room
# ======================================================================
def modifyClassRoom(id_room, name_room, capacity_room, wired_room, blackboard, address_room):   
    query = '''
            PREFIX uni: <http://www.rdfproject.com/>
            PREFIX un: <http://www.w3.org/2007/ont/unit#>
            WITH <http://www.rdcproject.com/graph/classrooms>
            DELETE { 
            ?x a uni:Classroom;
                uni:idRoom ?oldidRoom;
                uni:classroomname ?oldclassroomname;
                uni:address ?oldaddress;
                uni:capacity ?oldcapacity;
                uni:blackboard ?oldblackboard;
                uni:wired ?oldwired;
            }
            INSERT {  
            ?x a uni:Classroom;
                uni:idRoom "'''+id_room+'''";
                uni:classroomname "'''+name_room+'''";
                uni:address "'''+address_room+'''";
                uni:capacity "'''+capacity_room+'''";
                uni:blackboard "'''+blackboard+'''";
                uni:wired "'''+wired_room+'''";
            }
            WHERE { 
            ?x a uni:Classroom;
                uni:idRoom "'''+id_room+'''";
                OPTIONAL {
                    ?x a uni:Classroom;
                        uni:idRoom ?oldidRoom;
                        uni:classroomname ?oldclassroomname;
                        uni:address ?oldaddress;
                        uni:capacity ?oldcapacity;
                        uni:blackboard ?oldblackboard;
                        uni:wired ?oldwired;
                }
            }
            '''

    sparql.setQuery(query)
    sparql.setMethod('POST') 
    print query
    return sparqlQuery.query().convert()




# ======================================================================
# Insert Preference
# ======================================================================
def insertPreference(prof, Discipline6H, ConsecutiveDays, selectDay1, selectDay2, NoLessonAMPM,writeMethodRoom):
    graph_preferences = "http://www.rdcproject.com/graph/preferences"
    # Create a new Query
    query = '''
        PREFIX uni: <http://www.rdfproject.com/>
        PREFIX un: <http://www.w3.org/2007/ont/unit#>
        INSERT DATA
        { 
        GRAPH <'''+graph_preferences+'''>{
        uni:'''+ prof +''' a uni:Preference;
                                uni:sixHourSplit "'''+Discipline6H+'''"; 
                                uni:consecutiveDays "'''+ConsecutiveDays+'''"; 
                                uni:noLessonDay1 "'''+selectDay1+'''"; 
                                uni:noLessonDay2 "'''+selectDay2+'''"; 
                                uni:noLessonAMPM "'''+NoLessonAMPM+'''"; 
                                uni:isPreferenceOf uni:'''+prof+'''; 
                                uni:writeMethodRoom "'''+writeMethodRoom+'''".                            
            }
        }
        '''
    # Run the query and print the result
    sparql.setQuery(query)
    sparql.setMethod('POST') 
    print query
    sparql.query()

# ======================================================================
# Update Preference
# ======================================================================
def modifyPreference(prof, Discipline6H, ConsecutiveDays, selectDay1, selectDay2, NoLessonAMPM,writeMethodRoom): 
    isPreferenceOf = "uni:isPreferenceOf uni:" + prof + ";"

    query = '''
            PREFIX uni: <http://www.rdfproject.com/>
            PREFIX un: <http://www.w3.org/2007/ont/unit#>
            WITH <http://www.rdcproject.com/graph/preferences>
            DELETE { 
            ?x a uni:Preference;
                    uni:isPreferenceOf ?oldisPreferenceOf;
                    uni:sixHourSplit ?oldsixHourSplit;
                    uni:consecutiveDays ?oldconsecutiveDays;
                    uni:noLessonDay1 ?oldnoLessonDay1;
                    uni:noLessonDay2 ?oldnoLessonDay2;
                    uni:noLessonAMPM ?oldnoLessonAMPM;
                    uni:writeMethodRoom ?oldwriteMethodRoom;
            }
            INSERT {  
            ?x a uni:Preference;
                    uni:sixHourSplit "'''+Discipline6H+'''"; 
                    uni:consecutiveDays "'''+ConsecutiveDays+'''"; 
                    uni:noLessonDay1 "'''+selectDay1+'''"; 
                    uni:noLessonDay2 "'''+selectDay2+'''"; 
                    uni:noLessonAMPM "'''+NoLessonAMPM+'''"; 
                    '''+isPreferenceOf+'''
                    uni:writeMethodRoom "'''+writeMethodRoom+'''";
            }
            WHERE { 
            ?x a uni:Preference;
                uni:isPreferenceOf ? uni:''' + prof + ''';
                OPTIONAL {
                    ?x a uni:Preference;
                        uni:isPreferenceOf ?oldisPreferenceOf;
                        uni:sixHourSplit ?oldsixHourSplit;
                        uni:consecutiveDays ?oldconsecutiveDays;
                        uni:noLessonDay1 ?oldnoLessonDay1;
                        uni:noLessonDay2 ?oldnoLessonDay2;
                        uni:noLessonAMPM ?oldnoLessonAMPM;
                        uni:writeMethodRoom ?oldwriteMethodRoom;
                }
            }
            '''

    sparql.setQuery(query)
    sparql.setMethod('POST') 
    print query
    sparql.query()

# ======================================================================
# Delete Preference
# ======================================================================
def cancelPreference(prof):  
    query = '''
                PREFIX uni: <http://www.rdfproject.com/>
                PREFIX un: <http://www.w3.org/2007/ont/unit#>
                DELETE WHERE { 
                GRAPH <http://www.rdcproject.com/graph/preferences> {
                        ?object uni:isPreferenceOf uni:'''+prof+''';
                         ?property  ?value 
                }
                }
                '''

    sparql.setQuery(query)
    sparql.setMethod('POST') 
    print query
    sparql.query()

File: js/query.js
/**
 * ===========================================================================
 * File: query.js 
 * Author: Antonio Faienza, Luca Angelucci, Alessio Ciarrocchi
 * This file make a query from db for create a different Entities
 * ===========================================================================
 */


/***********************************
 * Query Professors
 ************************************/
function queryProfessors(returnValue, callback) {

    var endpointURL = "http://localhost:3030/ds/query";

    var myquery = ` PREFIX uni: <http://www.rdfproject.com/>
                    PREFIX un: <http://www.w3.org/2007/ont/unit#>

                    SELECT ?id ?name ?surname ?role
                    FROM <http://www.rdcproject.com/graph/professor>
                    WHERE
                    { ?x  a uni:Teacher.
                        ?x uni:idProfessor ?id.
                        ?x uni:firstName ?name.
                        ?x uni:lastName ?surname.
                        ?x uni:role ?role.                        
                        }
                    `;

    var encodedquery = encodeURIComponent(myquery);


    $.ajax({
        dataType: "jsonp",
        url: endpointURL + "?query=" + encodedquery + "&format=" + "json",
        success: function (results) {
            $.each(results, function (index, element) {
                var result = [];
                var bindings = element.bindings;
                // REF: https://www.w3.org/TR/rdf-sparql-json-res/
                for (i in bindings) {
                    var name = bindings[i].name.value;
                    var surname = bindings[i].surname.value;
                    var id = bindings[i].id.value;
                    var role = bindings[i].role.value;

                    var professor = new Professor(name, surname, id, role);
                    result.push(professor);
                }

                callback(result);

            });

        }

    });


}


/***********************************
 * Query Preferences
 ************************************/
function queryPreferences(returnValue, callback) {

    var endpointURL = "http://localhost:3030/ds/query";

    var myquery = ` PREFIX uni: <http://www.rdfproject.com/>           

                    SELECT DISTINCT ?idProf ?consecutiveDays ?setperiodofday ?noLessonDay1 ?noLessonDay2 
                                    ?sixHourSplit ?writeMethodRoom ?role

                    FROM <http://www.rdcproject.com/graph/disciplines>
                    FROM <http://www.rdcproject.com/graph/professor>    
                    FROM <http://www.rdcproject.com/graph/preferences>

                    WHERE
                    {                           
                        ?x  a uni:Discipline;
                            uni:isTaughtBy ?isTaughtBy.
                                ?isTaughtBy a uni:Teacher;
                            uni:role ?role;
                            uni:idProfessor ?idProf. 
                            ?isTaughtBy a uni:Preference;                             
                                uni:consecutiveDays ?consecutiveDays;
                                uni:noLessonAMPM ?setperiodofday;
                                uni:noLessonDay1 ?noLessonDay1;
                                uni:noLessonDay2 ?noLessonDay2;
                                uni:sixHourSplit ?sixHourSplit;
                                uni:writeMethodRoom ?writeMethodRoom.
                }ORDER BY (?role)
                    `;

    var encodedquery = encodeURIComponent(myquery);


    $.ajax({
        dataType: "jsonp",
        url: endpointURL + "?query=" + encodedquery + "&format=" + "json",
        success: function (results) {
            $.each(results, function (index, element) {
                var result = [];
                var bindings = element.bindings;
                // REF: https://www.w3.org/TR/rdf-sparql-json-res/
                for (i in bindings) {
                    
                    var idProf = bindings[i].idProf.value;
                    var consecutiveDays = bindings[i].consecutiveDays.value;
                    var setperiodofday = bindings[i].setperiodofday.value;
                    var noLessonDay1 = bindings[i].noLessonDay1.value;
                    var noLessonDay2 = bindings[i].noLessonDay2.value;
                    var sixHourSplit = bindings[i].sixHourSplit.value;
                    var writeMethodRoom = bindings[i].writeMethodRoom.value;
                                    
                    
                    result.push({
                        "idProf": idProf,
                        "consecutiveday": consecutiveDays,
                        "setperiodofday": setperiodofday,
                        "avoidlessonday1": noLessonDay1,
                        "avoidlessonday2": noLessonDay2,
                        "splitdurationlessons6h": sixHourSplit,
                        "blackboard": writeMethodRoom
                    });
                    
                }

                callback(result);

            });

        }

    });


}

/***********************************
 * Query Classes
 ************************************/
function queryClassrooms(returnValue, callback) {

    var endpointURL = "http://localhost:3030/ds/query";

    var myquery = ` PREFIX uni: <http://www.rdfproject.com/>
                    PREFIX un: <http://www.w3.org/2007/ont/unit#>

                    SELECT  ?id ?classroomname ?address ?capacity ?blackboard ?wired
                    FROM <http://www.rdcproject.com/graph/classrooms>
                    WHERE
                    { ?x  a uni:Classroom.
                        ?x uni:idRoom ?id.
                        ?x uni:classroomname ?classroomname.
                        ?x uni:address ?address.
                        ?x uni:capacity ?capacity.
                        ?x uni:blackboard ?blackboard.
                        ?x uni:wired ?wired.
                        }
                    `;

    var encodedquery = encodeURIComponent(myquery);


    $.ajax({
        dataType: "jsonp",
        url: endpointURL + "?query=" + encodedquery + "&format=" + "json",
        success: function (results) {
            $.each(results, function (index, element) {
                var result = [];
                var bindings = element.bindings;
                // REF: https://www.w3.org/TR/rdf-sparql-json-res/
                for (i in bindings) {
                    var id = bindings[i].id.value;
                    var name = bindings[i].classroomname.value;
                    var address = bindings[i].address.value;
                    var capacity = bindings[i].capacity.value;
                    var blackboard = bindings[i].blackboard.value;
                    var wired = bindings[i].wired.value;

                    var room = new Classroom(id, name, address, parseInt(capacity), blackboard, wired);
                    result.push(room);
                }
                callback(result);

            });

        }

    });

}


/***********************************
 * Query Course
 ************************************/
function queryCourses(returnValue, callback) {

    var endpointURL = "http://localhost:3030/ds/query";

    var myquery = ` PREFIX uni: <http://www.rdfproject.com/>
                    PREFIX un: <http://www.w3.org/2007/ont/unit#>

                    SELECT  ?id ?courseName 
                    FROM <http://www.rdcproject.com/graph/course>
                    WHERE
                    { ?x  a uni:Course.
                        ?x uni:idCourse ?id.
                        ?x uni:courseName ?courseName.
                    }
                    `;

    var encodedquery = encodeURIComponent(myquery);


    $.ajax({
        dataType: "jsonp",
        url: endpointURL + "?query=" + encodedquery + "&format=" + "json",
        success: function (results) {
            $.each(results, function (index, element) {
                var result = [];
                var bindings = element.bindings;
                // REF: https://www.w3.org/TR/rdf-sparql-json-res/
                for (i in bindings) {
                    var id = bindings[i].id.value;
                    var courseName = bindings[i].courseName.value;
                    var course = new Course(id, courseName);
                    result.push(course);
                }
                callback(result);
            });

        }

    });

}


/***********************************
 * Query Professors teacher a subject
 * REF: https://stackoverflow.com/a/18214142/4700162
 * @method queryDiscipline
 ************************************/
function queryDiscipline(sem) {
    var endpointURL = "http://localhost:3030/ds/query";
    var myquery = `
                    PREFIX uni: <http://www.rdfproject.com/>
                            

                    SELECT  ?idDiscipline ?sigleDiscipline ?disciplineName ?cfu  ?idCourse ?idCourse2 ?obligatory ?curriculum ?semester ?totalhours ?weekhours ?year ?numStudents                            
                            (GROUP_CONCAT(DISTINCT ?prof_str;separator=",") AS ?professors) 
                            
                    FROM <http://www.rdcproject.com/graph/disciplines>
                    FROM <http://www.rdcproject.com/graph/professor>
                    FROM <http://www.rdcproject.com/graph/course>
                    FROM <http://www.rdcproject.com/graph/preferences>
                    WHERE
                    {                           
                            {      
                            ?x  a uni:Discipline;
                            uni:disciplinename ?disciplineName;
                            uni:idDiscipline ?idDiscipline;
                            uni:disciplineAbbreviation ?sigleDiscipline;
                            uni:cfu ?cfu;
                            uni:hasCourseof ?hasCourseof;
                            uni:obligatory ?obligatory;
                            uni:curriculum ?curriculum;
                            uni:semester ?semester;
                            uni:semester "`+ sem + `";
                            uni:totalhours ?totalhours;
                            uni:weekhours ?weekhours;
                            uni:year ?year;
                            uni:numStudents  ?numStudents;
                            uni:isTaughtBy ?isTaughtBy.
                                ?isTaughtBy a uni:Teacher;
                                uni:idProfessor ?idProf. 
                            ?hasCourseof a uni:Course;
                                uni:idCourse ?idCourse.                        
                    }
                    
                    

                    BIND(CONCAT(?idProf) AS ?prof_str)
                    

                    }GROUP BY ?idDiscipline ?sigleDiscipline ?disciplineName ?cfu ?idCourse ?idCourse2 ?obligatory ?curriculum ?semester 
                    ?totalhours ?weekhours ?year ?numStudents 		
                    `;




    var encodedquery = encodeURIComponent(myquery);

    return new Promise(function (resolve, reject) {
        $.ajax({
            dataType: "jsonp",
            url: endpointURL + "?query=" + encodedquery + "&format=" + "json",
            success: function (results) {
                $.each(results, function (index, element) {

                    var result = [];
                    var bindings = element.bindings;

                    for (i in bindings) {
                        var d = new Discipline(
                            bindings[i].idDiscipline.value,
                            bindings[i].sigleDiscipline.value,
                            bindings[i].disciplineName.value,
                            bindings[i].semester.value,
                            bindings[i].obligatory.value,
                            splitCurriculum(bindings[i].curriculum.value),
                            bindings[i].totalhours.value,
                            bindings[i].weekhours.value,
                            bindings[i].cfu.value,
                            bindings[i].year.value,
                            parseInt(bindings[i].numStudents.value)
                        );
                        d.setCourse(bindings[i].idCourse.value);
                        var idProfs = bindings[i].professors.value.split(",");
                        for (var p = 0; p < idProfs.length; p++) {
                            d.setProfessor(idProfs[p]);
                        }
     
                        result.push(d);
                    }
                    if (result.length > 0) {
                        resolve(result);
                    }

                });
            }

        });
    });



}






